### ISSUE TITLE
Experiments in jMetal 5

### ISSUE THREAD

**Issue by ajnebro** [jMetal/jMetal#135]

**Comment by ajnebro** [comment]
# The list of algorithms
As we want to compare a number of algorithms, we need a list with them. The algorithms must be properly configured, so in all the experiments (currently included)[https://github.com/jMetal/jMetal/tree/master/jmetal-exec/src/main/java/org/uma/jmetal/experiment] there is a `configureAlgorithmList()` method to create this list.

As every algorithm will be executed `N` per configuration, the output files must be numerated. This is one of reason of defining a (`TaggedAlgorithm` class)[https://github.com/jMetal/jMetal/blob/master/jmetal-core/src/main/java/org/uma/jmetal/util/experiment/util/TaggedAlgorithm.java]:
```java
  private Algorithm<Result> algorithm ;
  private Problem<?> problem ;
  private String tag ;
  private int runId;

  public TaggedAlgorithm (Algorithm<Result> algorithm, Problem<?> problem, int runId) {
    this(algorithm, algorithm.getName(), problem, runId) ;
  }

  public TaggedAlgorithm (Algorithm<Result> algorithm, String tag, Problem<?> problem, int runId) {
    this.algorithm = algorithm ;
    this.tag = tag ;
    this.problem = problem ;
    this.runId = runId ;
  }
```
This class contains all the information required to run a particular configuration: the algorithm, the problem, and the run indentifier. A tag is also needed to create a directory with the name of the algorithm. All the algorithms implements the [`DescribedEntity`](https://github.com/jMetal/jMetal/blob/master/jmetal-core/src/main/java/org/uma/jmetal/util/naming/DescribedEntity.java) interface, so they have a `getName()` method, which is used by default as tag. However, if we want to compare different versions of the same algorithm we need a different name for them; that's what the use of tags allows. An example of experiment of this type is the [NSGAIIStudy](https://github.com/jMetal/jMetal/blob/master/jmetal-exec/src/main/java/org/uma/jmetal/experiment/NSGAIIStudy.java).

# The list of problems
Currently, the list of algorithms is merely a list:
```java
List<Problem<DoubleSolution>> problemList = Arrays.<Problem<DoubleSolution>>asList(new ZDT1(), new ZDT2(),
        new ZDT3(), new ZDT4(), new ZDT6()) ;
```
As in the case of the algorithms, we could be interested in solving different variants of a same problem (for example, the `ZDT1` with 30, 100, 500, and 1000 decision variables in a scalability study). This is not possible now (and the reason of opening the #129 issue).

**Comment by ajnebro** [comment]
# Output of an experiment
After the successful execution of an experiment the results are structured in directory tree as shown in this example:

![outputdirectorystructure](https://cloud.githubusercontent.com/assets/1889196/13680638/ce7c471e-e6f9-11e5-9c8c-f249c9428ddf.png)

We can see that there is a `data` directory containing a subdirectoy per algorithm, and each of them contains a subdirectory per problem.

To be continued ....

**Comment by matthieu-vergne** [comment]
I have finally some time for me, so now I can think about it.

Thanks a lot for your effort, it is so clear I am currently working on an architecture on my side, and my objective is to come with something working where I have a translation of one of your experiments into my architecture, so not only I can check it works but you can also go progressively though it to see if you like it.

But here I am facing a first hard point: conceptually speaking, you are not experimenting a set of algorithms over a set of problems. You are experimenting a set of algorithms, period. Why? Because your algorithms depend on your problems, and so actually, you only need your problems to configure your algorithms. So at the end, you only experiment on a bunch of algorithms, some being set on problem X, others on problem Y, etc.

And this is giving me a hard time to represent a set of algorithms and a set of problems independently in my architecture, because you cannot establish "I want to try this algorithm configured in this way" without establishing the problem first, so once you reach the level of the experiment, you just don't care about the problem anymore.

So here is my point: would you like to make your algorithms independent from your problems? If yes, then I can work on that aspect first to suggest you some ways to do it, and then come back on the experiments. If not, then I should assume that you plan to revise your description of experiment as a way to compare algorithms, and forget about problems at this level. What is your position?

I think you already know mine: I would prefer them to be independent.

**Comment by Juanjdurillo** [comment]
Although I've not been active in the discussions, I am willing to know more about your proposal. So far it's hard for me not to have the same view as Antoni: dependence problem-algorithm

**Comment by ajnebro** [comment]
> problems are part of the algorithm setting

This is the point. Problem and algorithm are independent entities, but when you use the latter to solve the former then the algorithm settings depend on the problem.

**Comment by matthieu-vergne** [comment]
I basically see two alternatives, which one would you support?

1) An algorithm is problem dependent:
```
algorithm = new MyAlgo(problem, param1, ..., paramN);
...
algorithm.run();
```

2) An algorithm is problem independent, but can be applied on a problem to solve it:
```
algorithm = new MyAlgo(param1, ..., paramN);
...
algorithm.run(problem);
```

**Comment by matthieu-vergne** [comment]
The 1st point implies that, to solve another problem, you have to make a new instance, which is why the problem "is part of the setting" (it is not really true, but from a usage perspective, this is how it looks like). The second allows you to set up everything, and at the last moment to decide on which problem you apply it, which means that nothing forbids you to apply the same algorithm instance to another problem just after.

**Comment by ajnebro** [comment]
The fast answer would be "alternative 2), of course", but this option wouldn't allow to get information of the problem in the algorithm class constructor.

**Comment by ajnebro** [comment]
Another point to take into accont is that an experiment typically envolves executing thousands of algorithm runs, so at least parallelism at thread level is a must. This implies that both the algorithm and the problem codes must be thread-safe if we want to use only one instance of them. As this is not the general case, the adopted approach is to have a list with all the tasks to be executed. 

My first approach was to the create an instance of each algorithm with the proper settings (including the problem), resulting in a list of A * P components (A algorithms, P problems). Then, to use the thread pool and executor features of Java, the list was traversed and, for each of its components, a task list of N elements (N independent runs) was created; each task contained the same algorithm instance. This was a wrong solution due to the aforementioned thread-safe issue. 

To avoid this problem my second approach was to clone the algorithm objects prior to include them into the tasks (this way, each task had a different instance of the same algorithm object). I used the `SerializationUtils.clone()` method of Apache Commons, which requires that everything must serialized, what was a headache (remember issue #130). 

Third and current approach is to create the N * A * P algorithm in the class defining the experiment.

**Comment by Juanjdurillo** [comment]
The point is whether that information is needed in the class constructor. If it is needed, then the dependence is justified in my view. So far I do not remember any case (besides the probability for mutation) when a constructor needed problem info. Do you remember any other case Antonio?

**Comment by ajnebro** [comment]
I made me that question a few minutes ago. 

Some examples: AbYSS,  GWASFGA, AbstractMOEAD, NSGAIII, SMPSO.

**Comment by Juanjdurillo** [comment]
well, looking at NSGAIII there is nothing that really needs to be in the constructor ... Actually only the number of objectives, which are needed for computing the reference points. But the reference point computations can also be moved to the run methods (do they need to be part of the constructor?)

**Comment by ajnebro** [comment]
The issue is that if we remove the problem-dependant code from the constructor, where do we put it?

NSGA-III and most of algorithms implements a template that does not have any method for initialization. We should then include something like an `init()` method in all the templates ... I'm not sure to like that idea ...

**Comment by ajnebro** [comment]
> The 1st point implies that, to solve another problem, you have to make a new instance, which is why the problem "is part of the setting" (it is not really true, but from a usage perspective, this is how it looks like). 

We have to consider also that we usually use builder classes to set the algorithm parameters, and the first parameter of the builder constructors is the problem.

**Comment by matthieu-vergne** [comment]
Please discuss the dependency stuff on #55, which is exactly for that. {^_^}

**Comment by ajnebro** [comment]
# Computing quality indicators and generating Latex tables from external data
We have commented before that the steps to carry out an experimental study are these:
```java
    new ExecuteAlgorithms<>(experiment).run();
    new GenerateReferenceParetoSetAndFrontFromDoubleSolutions(experiment).run();
    new ComputeQualityIndicators<>(experiment).run() ;
    new GenerateLatexTablesWithStatistics(experiment).run() ;
    new GenerateWilcoxonTestTablesWithR<>(experiment).run() ;
    new GenerateFriedmanTestTables<>(experiment).run();
    new GenerateBoxplotsWithR<>(experiment).setRows(3).setColumns(3).setDisplayNotch().run() ;
```

However, there are situations where the `ExecuteAlgorithms` step is not used. Two of those situations are:
* Executing the algorithms takes too much time in a single multi-core processor. In that case, we usually use a Condor cluster to use hundreds of cores. 
* We are using jMetalCpp, which does not include support for experiments.

In these cases, what we do is to arrange all the result data (FUNxx and VARxx files) according to the expected output directory structure as if they were been obtained with `ExecuteAlgorithms` and then the rest of the code is executed.

**Comment by matthieu-vergne** [comment]
Well, there is actually several drawbacks in the code you mention:
- you mix everything: data generation, data storage, results computation, LaTeX rendering, etc.
- the flow is completely hidden: you give always the same `experiment` but never get anything from it, so I assume it somehow store everything or something like that.

In [my new architecture](https://github.com/matthieu-vergne/jMetal/tree/experiment/jmetal-experiment/src/main/java/org/uma/jmetal/experiment), the first point is solved by having dedicated classes. So you have an `ExperimentFeeder` which decides what are the algorithm/problem combinations to run, an `ExperimentExecution` (I should call it executor...) to manage the execution (including parallelism), and one or several `ExperimentRenderer` instances, each dedicated to obtaining specific results (through getters/listeners) and rendering them in a specific way (e.g. one for data storage, one for LaTeX tables, etc.).

The second point is solved by using a clear flow:
- the feeder is just an `ITerable` container, so passive,
- the execution is the active one, so it reads the feeder and has getters and listeners
- the renderers read the getters and register to the listeners to obtain the data they need and render it the way they want

An advantage is that nothing needs to be stored a priori: the use of listeners allows to provide the data immediately to the renderers needing it, while it is garbage-collected if no one need it or as soon as it has been used. Another advantage is the modularity: if you don't need a full experiment, then just take the parts you need. If your data is already generated, you could have some kind of `RetrospectiveExperimentExecution` which just reads some data stored and fake an execution by providing all the data stored at the rythm it reads it. The key here is that all the data *must* be generated by the execution part, not the renderer. So the renderer should not even measure time or stuff like that. I think you can already implement such kind of dummy with your architecture, by the way, although I have no idea how complex it would be.

**Comment by matthieu-vergne** [comment]
A question about the current implementation: you have `ExperimentAlgorithm` and `ExperimentProblem` to manage tags. Now, if I look at the implementation of `GenerateReferenceParetoFront`, sometimes you use the tags, sometimes you use the names. In particular, you use tags everywhere but there:
```java
new SolutionListOutput(solutionsPerAlgorithm)
  .printObjectivesToFile(
    outputDirectoryName + "/" + problem.getName() + "." +
    algorithm.getAlgorithmTag() + ".rf"
  );
```

Is it a mistake, and should it use tags everywhere, or is it on purpose?

**Comment by ajnebro** [comment]
It is a mistake: it should be `problem.getTag()` instead of `problem.getName()`. 

The point here is that algorithms and problems have a `getName()` method, but if we want to use several instances of a same algorithm (each one with different settigs) or of the same problem we cannot use `getName()`.

**Comment by matthieu-vergne** [comment]
I thought so.

Anyway, the framework is DONE! Well, only the basis. There is still plenty of things we could do, but at least I was able to make a complete experiment. There is no test so all I did was run it to see if it logs similar stuff without throwing exceptions. The framework really is about the design of the experiment, not low level execution (I reused stuff), so I did not check the content of the files, which should be the same. My last question for example was about a specific component but I just reuse it as is. It can be fixed later.

I could present it in various ways, but because you are the concrete one, I will start by showing you how it looks like.

The experiment I translated is `BinaryProblemsStudy`. No particular reason for choosing this one beside being the first in the list. It has as main code this piece:
```java
List<ExperimentProblem<BinarySolution>> problemList = new ArrayList<>();
problemList.add(new ExperimentProblem<>(new ZDT5()));
problemList.add(new ExperimentProblem<>(new OneZeroMax(512)));

List<ExperimentAlgorithm<BinarySolution>> algorithmList =
		configureAlgorithmList(problemList);

int numberOfCores = 8;
List<GenericIndicator<BinarySolution>> indicatorList = Arrays.asList(
		new Epsilon<BinarySolution>(), new Spread<BinarySolution>(), new GenerationalDistance<BinarySolution>(),
		new PISAHypervolume<BinarySolution>(),
		new InvertedGenerationalDistance<BinarySolution>(),
		new InvertedGenerationalDistancePlus<BinarySolution>()) ;
String referenceFrontDirectory = experimentBaseDirectory+"/referenceFronts" ;
String outputParetoFrontFileName = "FUN" ;
String outputParetoSetFileName = "VAR" ;
String experimentName = "BinaryProblemsStudy" ;

new ExecuteAlgorithms<>(algorithmList, INDEPENDENT_RUNS, numberOfCores, experimentBaseDirectory).run();
List<String> referenceFrontFileNames = new LinkedList<>();
new GenerateReferenceParetoFront(algorithmList, problemList, experimentBaseDirectory, referenceFrontDirectory, outputParetoFrontFileName, INDEPENDENT_RUNS, referenceFrontFileNames).run();
new ComputeQualityIndicators<>(algorithmList, problemList, indicatorList, experimentBaseDirectory, referenceFrontDirectory, referenceFrontFileNames, outputParetoFrontFileName, outputParetoSetFileName, INDEPENDENT_RUNS).run();
new GenerateLatexTablesWithStatistics(algorithmList, problemList, indicatorList, experimentBaseDirectory, experimentName).run();
new GenerateWilcoxonTestTablesWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
new GenerateFriedmanTestTables<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
new GenerateBoxplotsWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).setRows(1).setColumns(2).setDisplayNotch().run();
```

I will start with a global overview and then go through each piece.

# Overall Structure

First of all, the current experiment is structured as such:
1. Specify the stuff to run
	a. Specify the problems to experiment on (`List<ExperimentProblem<BinarySolution>> problemList`), here `ZDT5` and `OneZeroMax` with 512 bits.
	b. Specify the algorithms to use to solve these problems (`List<ExperimentAlgorithm<BinarySolution>> algorithmList`), here `NSGAII`, `SPEA2`, `MOCell`, and `MOCHC` if I rely on the builders you use.
	c. Specify the number of times each (algorithm, problem) couple should be run (`int INDEPENDENT_RUNS = 25`). For this one, you spend some efforts instantiating all the algorithms, for each problem, one time for each run. Clearly, this is something to be done for each experiment, and thus something I would expect to be managed by the experiment itself, not manually. It also forces you to remove the run duplicates later, for components which care about the (algorithm, problem) couples rather than for each run specifically. If we don't manage the instantiation manually, this is an inconvenient we don't face anymore, as long as we can still obtain the couples.
2. Execute the algorithms (`new ExecuteAlgorithms<>(...).run()`) to produce raw data (e.g. FUN files, VAR files). You made it like a mere component to be executed, but this is a fundamental part of an experiment. Without execution and data production, there is no experiment at all, while we could imagine experiments without Pareto fronts nor Wilcoxon tests. So rather than an additional component, it should be a default, with the possibility to configure it based on our needs (e.g. sequential computation, thread-based computation, external computation).
3. Refine the raw data into other intermediary resources (e.g. reference Pareto fronts, quality indicators) and exploitable results (e.g. tables, scripts).

In the framework, I follow the same process. But unlike your code, which execute pieces one after the other, I use an experiment builder to specify everything I need. Only at the end, the actual experiment is built and run. So all the code I will present is enclosed in that:
```java
PreparePerformProduceExperimentBuilder<Algorithm<List<BinarySolution>>, BinaryProblem> experimentBuilder = new PreparePerformProduceExperimentBuilder<>();

// Experiment building...

Experiment experiment = experimentBuilder.build();
experiment.run();
```

As you may guess, an `Experiment` is nothing more than a `Runnable` overriden, with its own documentation. This is the generic structure of an experiment, whatever its purpose. What makes the specific organization is the builder: the builder is designed to create a specific kind of experiment, here a `PreparePerformProduceExperiment`. This name reflects the 3 main steps of the procedure above: specify the stuff to run (prepare), run to produce data (perform), and produce exploitable results out of the raw data (produce). We may call it P<sup>3</sup> (p-cube) experiment, if you want a catchy name. This experiment is configured here to solve `BinaryProblem` instances with any `Algorithm`.

Now, let's see how each phase has been translated for using the framework.

# Specify the stuff to run

All this part is about preparing the algorithms and problems to run. For that, we need to specify which problems to solve, which algorithms to use for that, and how many times to run them. All of them relate to the first phase (prepare) and thus are done through methods having this pattern `experimentBuilder.prepareXxx(...)`.

## Specify the problems to experiment on

Here is your code:
```java
List<ExperimentProblem<BinarySolution>> problemList = new ArrayList<>();
problemList.add(new ExperimentProblem<>(new ZDT5()));
problemList.add(new ExperimentProblem<>(new OneZeroMax(512)));
```

Here is mine:
```java
List<ProblemID<? extends BinaryProblem>> problemIDs = new LinkedList<>();
problemIDs.add(experimentBuilder.prepareProblem(() -> new ZDT5()));
problemIDs.add(experimentBuilder.prepareProblem(() -> new OneZeroMax(512)));
```

Clearly, we do the same thing, but with some significant differences:
- we do not instantiate the problems, we only tell how to do so (notice the Lambda expressions). This is the builder which will do it for us.
- The builder does not return an instance of the problem but a `ProblemID`. This is through this ID that we will access the problem instance later when needed.

## Specify the algorithms to use to solve these problems

You factored this part into a dedicated method. If I dig in it (and remove the details for readability), here is your code:
```java
List<ExperimentAlgorithm<BinarySolution>> algorithms = new ArrayList<>();

for (int i = 0; i < problemList.size(); i++) {
	Algorithm<List<BinarySolution>> algorithm = new NSGAIIBuilder<BinarySolution>(...) ... .build();
	algorithms.add(new ExperimentAlgorithm<>(algorithm, problemList.get(i).getTag()));
}

for (int i = 0; i < problemList.size(); i++) {
	Algorithm<List<BinarySolution>> algorithm = new SPEA2Builder<BinarySolution>(...) ... .build();
	algorithms.add(new ExperimentAlgorithm<>(algorithm, problemList.get(i).getTag()));
}

for (int i = 0; i < problemList.size(); i++) {
	Algorithm<List<BinarySolution>> algorithm = new MOCellBuilder<BinarySolution>(...) ... .build();
	algorithms.add(new ExperimentAlgorithm<>(algorithm, problemList.get(i).getTag()));
}

for (int i = 0; i < problemList.size(); i++) {
	...
	Algorithm<List<BinarySolution>> algorithm = new MOCHCBuilder(...) ... .build();
	algorithms.add(new ExperimentAlgorithm<>(algorithm, problemList.get(i).getTag()));
}
```

Here is mine:
```java
List<AlgorithmID<? extends Algorithm<List<BinarySolution>>>> algorithmIDs = new LinkedList<>();
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new NSGAIIBuilder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new SPEA2Builder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new MOCellBuilder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> {
		...
		return new MOCHCBuilder(...) ... .build();
	}));
```

Similarly to the problems:
- we do not instantiate the algorithms, we only tell how to do so (notice the first Lambda expressions for each). This is the builder which will do it for us.
- The builder does not return an instance of the algorithm but an `AlgorithmID`. This is through this ID that we will access the algorithm instance later when needed.

Additionally, we don't care yet about how many times they will be run. This is something that we let for the next step. Sincerely, because of the pending issues about algorithms, problems, and solutions (e.g. #55, #71), I tried to keep it highly generic first. It is only recently that I bounded it to `Algorithm` to reuse the few facilities it provides. It is not bounded to `Problem`, however, because nothing of it is reused in the builder. So P<sup>3</sup> experiments can be used to solve any problem, not necessarily jMetal ones, and the `Algorithm` interface is also simple enough to be easy to use on external implementations, so it is not really a limitation either.

## Specify the number of times each (algorithm, problem) couple should be run

Your code for that is decomposed in 2 steps:
1. defining the `INDEPENDENT_RUNS` constant
2. instantiating each algorithm through `for` loops, which you did in the previous step

In my code, I did not made the `for` loops in the previous step, but I won't do them here either. My code is thus reduced to this:
```java
experimentBuilder.prepareIndependentRuns(INDEPENDENT_RUNS);
```

When the builder will instantiate the experiment at the end, this is when all the algorithm and problem instances will be produced based on what we gave before, including for each independent run. So all what we need to do now is to say how many times to run them. Meanwhile, this property has a default value of 1, like in your own `Experiment` class.

# Execute the algorithms to produce raw data

So far, all the *prepare* phase has been specified. The next phase is *perform*, which aims at executing each (algorithm, problem) couple with N independent runs. For this phase, there is 2 things to do. One is to decide how to execute them, and the second is to specify all the raw data we want to produce during this execution. Your code do all of that through `new ExecuteAlgorithms<>(...).run()`, but the point is that it hides everything, so I extracted the code of this component to show everything. Of course it can be hidden again by factoring the code, but the stuff created here should be used in the next components to produce tables and so on, so let's show it for now.

## Algorithm execution

This is the core of the *perform* phase, and thus it is already there by default. Initially, it will just execute the runs sequentially, which makes the log easier to read, but this behaviour can be changed by using this method:
```java
experimentBuilder.performWith(new ThreadPerformer(8));
```

`ThreadPerformer` is an executor which uses a pool of threads (with the help of `Executors`) to run the algorithms in parallel, based on the number of cores provided. You used a parallel execution too, so I did that. You pass through the collection stream, which is possible since Java 8, while I pass through an `ExecutorService`, which is there since Java 5. You need to specify a `System` property to set the cores, I only need to provide the number of cores as a parameter. Meanwhile, if you don't explicit the number of cores, it will take as default `Runtime.getRuntime().availableProcessors()`.

The interface to implement is obscenely simple:
```java
public static interface Performer {
	public void request(Runnable runnable);
	public void runUntilAllRequestsAreTerminated();
}
```

So it is really easy to make new execution strategies. There is plenty of interfaces and classes created by this builder, but they all remain extremely simple, so as long as the purpose of each method is understood, it should be straightforward to use. I won't show each of them now, to remain sho... hum... not too long, but simplicity is a general approach, in terms of number of methods as well as number of parameters.

## Raw data production

The previous stuff is all what you need to have a valid experiment. You can then instantiate it and run it. But usually, you make an experiment for producing specific results, like here tables and scripts. The architecture chosen here splits the execution and the production of results into 2 separate steps (*perform* and *produce*). The idea is that the execution should be as much as possible free from extra resource consumption, in order to have a reliable evaluation of the performances. As such, the execution should only produce raw data which minimises such an impact. All the extra computation and rich formatting should be delegated to the last step, which we will see later.

In your case, the raw data you produce are the FUN and VAR files. This is what is done in your `ExecuteAlgorithms`, and so this is what I took as raw data for my framework. But to create them you use various other pieces of data. The point is that these pieces of data are specific to each run. Let's go progressively to see how we do that with the framework.

You define a piece of data like this:
```java
DataID<String> problemTag = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getProblem().getName());
```

This line shows several things:
- you *define* a piece of data, you don't generate it. It will be generated on demand if required.
- if it is requested, it will be through the ID returned by the method, `problemTag` here.
- you say whether or not it should be stored for reuse (`ONCE_AND_BACKUP`) or genereted every time we request it (`EVERY_CALL`).
- you say how it should be produced, which is a *contextual* operation, and you can access this context through the parameter given (named `context` here).

The context here is the run: the experiment runs algorithms separately, and this raw data relates to each run separately. As such, you can generate common data specific to each run, here a problem tag which relates to the problem solved during the run. With such a definition, each run will have its own problem tag based on the name of its own problem instance. The runs solving the same kind of problem will have the same tag content, but they will have their own tag instance, like they have their own problem instance. So it is thread safe.

The context object provides several information related to the current run:
- the algorithm executed (`context.getAlgorithm()`)
- the problem solved, which is used here (`context.getProblem()`)
- the run index (`context.getRun()`)
- an access to other run-specific data (`context.getData(id)`)

The 3 firsts are straightforward, they relate to what identifies the current run. Notice that, because we are bound to `Algorithm`, you can always access the current result through `context.getAlgorithm().getResult()`. The last one gives you access to other pieces of data you may define. For example, here we have defined a `problemTag`. We can define similarly an `algorithmTag`. Once we have both of them, we can reuse them to define another piece of data like this one:
```java
DataID<String> outputDirectoryName = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> experimentBaseDirectory + "/data/" + context.getData(algorithmTag) + "/" + context.getData(problemTag));
```

If this new data is requested, it will first request the `algorithmTag` of this run, which will be generated if it has not been yet or retrieved from a backup, then request the `problemTag` of this run similarly, and then produce the `outputDirectoryName`, which will be also stored for reuse. If we consider all the definitions made for this experiment, this is what I have:
```java
DataID<String> problemTag = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getProblem().getName());
DataID<String> algorithmTag = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getAlgorithm().getName());
DataID<String> outputDirectoryName = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> experimentBaseDirectory + "/data/" + context.getData(algorithmTag) + "/" + context.getData(problemTag));
DataID<String> funFile = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getData(outputDirectoryName) + "/FUN" + context.getRun() + ".tsv");
DataID<String> varFile = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getData(outputDirectoryName) + "/VAR" + context.getRun() + ".tsv");
```

Here you see that we do not exploit a whole tag algorithm+problem+run, but the three elements are indeed used to produce files... Well, actually to produce the paths of these files. The point is that we did not produce anything yet, we only defined different pieces of data which could be produced during the run. What remains now is to tell when to produce them and how to use them to produce the FUN and VAR files.

Your `ExecuteAlgorithms` is organised to run algorithms wrapped into `ExperimentAlgorithm` instances. If we check the code of this stuff, here is what it does:
```java
String outputDirectoryName = experimentBaseDirectory
		+ "/data/"
		+ algorithmTag
		+ "/"
		+ problemTag;

File outputDirectory = new File(outputDirectoryName);
if (!outputDirectory.exists()) {
	boolean result = new File(outputDirectoryName).mkdirs();
	if (result) {
		JMetalLogger.logger.info("Creating " + outputDirectoryName);
	} else {
		JMetalLogger.logger.severe("Creating " + outputDirectoryName + " failed");
	}
}

String funFile = outputDirectoryName + "/FUN" + id + ".tsv";
String varFile = outputDirectoryName + "/VAR" + id + ".tsv";
JMetalLogger.logger.info(
		" Running algorithm: " + algorithmTag +
				", problem: " + problemTag +
				", run: " + id +
				", funFile: " + funFile);


algorithm.run();
List<? extends S> population = algorithm.getResult();

new SolutionListOutput(population)
		.setSeparator("\t")
		.setVarFileOutputContext(new DefaultFileOutputContext(varFile))
		.setFunFileOutputContext(new DefaultFileOutputContext(funFile))
		.print();
```

More informally:
1. Create folders and variables, to some extent this is what we did by defining our data (the folders are not created yet, but all the run-specific paths are ready to be called)
2. Log that we will run an algorithm
3. Run the algorithm
4. Create the FUN and VAR files

In other words, we have the algorithm run, some stuff to do before, and some stuff to do after. With our experiment builder, the run is automatic, but we can specify events to be executed before and after the run. I have changed a bit the code I extracted, but this is more or less the same thing, just don't care about the small differences. Before the run, we need to log that a specific algorithm will be run, which is done that way:
```java
experimentBuilder.createRunEvent(When.BEFORE_RUN, (context) ->
	JMetalLogger.logger.info(
		"Running algorithm: " + context.getAlgorithm().getName() +
				", problem: " + context.getProblem().getName() +
				", run: " + context.getRun() +
				", funFile: " + context.getData(funFile) +
				", varFile: " + context.getData(varFile)));
```
I don't create the folders now because you create the files only once the algorithm is finished, so I moved it to after the run. As you can see, this is again something simple: you create an event which consists in doing something (here logging) and you say when it should be executed, here `BEFORE_RUN`. This means that, before to automatically run the algorithm, this piece of code will be executed. You can also see that it receives the `context` we used when defining our data, so you can reuse it to access all the data you have defined before, as well as the algorithm, problem, and run index, which is what we do here for logging. This is because these events are executed that the data you defined before will be generated. A piece of data that you define but never call here or later will never be generated.

You can define several events like that, but they are independent, so you cannot guaranty an order between them. Normally, you specify only 1 such event and put everything you need inside. It is nevertheless possible to create others, which can be useful for example if you want to add a predefined piece. We can imagine for example that you have a predefined package, for instance all your experiments reuse several components, this is an exemple of package to be made. You apply this package through a single method call, which specifies its own data and pre-/post-run events. If you want to add your own, you can do that independently of what the package has already done. If the package gives you the `DataID` instances it has produced, you can also reuse its own data. But that is for more advanced experiment design. Let's focus on our example.

Once we have specified what should be done *before* to run the algorithm, we can specify what should be done *after*, which is nothing different, just change `BEFORE_RUN` by `AFTER_RUN` and give the relevant content:
```java
experimentBuilder.createRunEvent(When.AFTER_RUN, (context) -> {
	JMetalLogger.logger.info(
		"Finished algorithm: " + context.getAlgorithm().getName() +
				", problem: " + context.getProblem().getName() +
				", run: " + context.getRun());
	new File(context.getData(outputDirectoryName)).mkdirs();
	
	List<BinarySolution> population = context.getAlgorithm().getResult();
	new SolutionListOutput(population)
		.setSeparator("\t")
		.setVarFileOutputContext(new DefaultFileOutputContext(context.getData(varFile)))
		.setFunFileOutputContext(new DefaultFileOutputContext(context.getData(funFile)))
		.print();
	JMetalLogger.logger.info(
		"Generated funFile: " + context.getData(funFile) +
				", varFile: " + context.getData(varFile));
});
```
Here we create the folders and the files inside them. I also added a bit of logging for clarity. Once again, it relates to a specific run, so you can use the `context` to access run-specific data, in particular to retrieve the directory and files to write into. Notice that you obtain the population by getting the algorithm from the context, so no need for an additional method. Now, each time an algorithm is executed, this piece of code will be executed just after, thus ensuring that we have the raw data that we need. Instead of using files, we could have exploited the context to store the data internally by defining data with the content instead of only the file path.

# Refine the raw data into other intermediary resources and exploitable results

The final phase is *produce*, where we refine the raw data to obtain exploitable results. In your code, you have a sequence of components to be run. I have shown in issue #221 how I refined them to use explicit parameters, which is what ensures that we execute them in the right order. In the following, I create all the parameters together for readability. So we need to translate something like this:
```java
// Create explicit parameters
int numberOfCores = 8;
List<GenericIndicator<BinarySolution>> indicatorList = Arrays.asList(
		new Epsilon<BinarySolution>(), new Spread<BinarySolution>(), new GenerationalDistance<BinarySolution>(),
		new PISAHypervolume<BinarySolution>(),
		new InvertedGenerationalDistance<BinarySolution>(),
		new InvertedGenerationalDistancePlus<BinarySolution>()) ;
String referenceFrontDirectory = experimentBaseDirectory+"/referenceFronts" ;
String outputParetoFrontFileName = "FUN" ;
String outputParetoSetFileName = "VAR" ;
String experimentName = "BinaryProblemsStudy" ;
List<String> referenceFrontFileNames = new LinkedList<>();

// Call the components in the right order
new GenerateReferenceParetoFront(algorithmList, problemList, experimentBaseDirectory, referenceFrontDirectory, outputParetoFrontFileName, INDEPENDENT_RUNS, referenceFrontFileNames).run();
new ComputeQualityIndicators<>(algorithmList, problemList, indicatorList, experimentBaseDirectory, referenceFrontDirectory, referenceFrontFileNames, outputParetoFrontFileName, outputParetoSetFileName, INDEPENDENT_RUNS).run();
new GenerateLatexTablesWithStatistics(algorithmList, problemList, indicatorList, experimentBaseDirectory, experimentName).run();
new GenerateWilcoxonTestTablesWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
new GenerateFriedmanTestTables<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
new GenerateBoxplotsWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).setRows(1).setColumns(2).setDisplayNotch().run();
```

In my code, this last phase has no particular constraint: the experimenter can use and do what he wants, the focus was on factoring the algorithm+problem instantiation and isolating execution from richer computation. Thus it is only a matter of calling `experimentBuilder.produce(...)`, which again takes a lambda which tells what should be done once all the algorithms have finished:
```java
experimentBuilder.produce((dataset) -> {
	try {
		// Retrieve the algorithms and problems, without duplicates
		List<ExperimentProblem<BinarySolution>> problemList = new LinkedList<>();
		List<ExperimentAlgorithm<BinarySolution>> algorithmList = new LinkedList<>();
		for (AlgorithmID<? extends Algorithm<List<BinarySolution>>> algorithmID : algorithmIDs) {
			for (ProblemID<? extends BinaryProblem> problemID : problemIDs) {
				RunDataset<? extends Algorithm<List<BinarySolution>>, ? extends BinaryProblem> context = dataset.getRunDataset(algorithmID, problemID, 0);
				problemList.add(new ExperimentProblem<>(context.getProblem(), context.getData(problemTag)));
				algorithmList.add(new ExperimentAlgorithm<>(context.getAlgorithm(), context.getData(algorithmTag), context.getData(problemTag)));
			}
		}
		
		// Create explicit parameters
		int numberOfCores = 8;
		List<GenericIndicator<BinarySolution>> indicatorList = Arrays.asList(
				new Epsilon<BinarySolution>(), new Spread<BinarySolution>(), new GenerationalDistance<BinarySolution>(),
				new PISAHypervolume<BinarySolution>(),
				new InvertedGenerationalDistance<BinarySolution>(),
				new InvertedGenerationalDistancePlus<BinarySolution>()) ;
		String referenceFrontDirectory = experimentBaseDirectory+"/referenceFronts" ;
		String outputParetoFrontFileName = "FUN" ;
		String outputParetoSetFileName = "VAR" ;
		String experimentName = "BinaryProblemsStudy" ;
		List<String> referenceFrontFileNames = new LinkedList<>();

		// Call the components in the right order
		new GenerateReferenceParetoFront(algorithmList, problemList, experimentBaseDirectory, referenceFrontDirectory, outputParetoFrontFileName, INDEPENDENT_RUNS, referenceFrontFileNames).run();
		new ComputeQualityIndicators<>(algorithmList, problemList, indicatorList, experimentBaseDirectory, referenceFrontDirectory, referenceFrontFileNames, outputParetoFrontFileName, outputParetoSetFileName, INDEPENDENT_RUNS).run();
		new GenerateLatexTablesWithStatistics(algorithmList, problemList, indicatorList, experimentBaseDirectory, experimentName).run();
		new GenerateWilcoxonTestTablesWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
		new GenerateFriedmanTestTables<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
		new GenerateBoxplotsWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).setRows(1).setColumns(2).setDisplayNotch().run();
	} catch (IOException cause) {
		throw new RuntimeException(cause);
	}
});
```

The first thing to notice is that, as opposed to run data and events, we won't access previous data through a `context` variable, but through a whole `dataset`. Indeed, while before we were defining operations to do for each run, motivating the use of a local context, here we speak about a single operation to do after everything has been run. Consequently, we obtain a dataset which provides everything. It gives access to the same elements, but with the additional requirement of specifying which run to access, so we have the following methods:
- `dataset.getAlgorithm(algorithmID, problemID, run)`
- `dataset.getProblem(algorithmID, problemID, run)`
- `dataset.getData(algorithmID, problemID, run, dataID)`

This way, everything that could be accessed during a specific run is also accessible during this phase. For readability, it is however more interesting to focus on a specific run with the method `dataset.getRunDataset(algorithmID, problemID, run)`, which returns a `RunDataset` that corresponds to the `context` variable that we used in the run-specific operations. This is by the way how these `context` variables are fed, by calling the method of the global dataset, so it is strictly the same (not the same instance, but the same content). Moreover, as the global `dataset` provides a local one, the local one also provides the global one through `getGlobalDataset()`. So even in the run-specific operations, it is possible to access the whole dataset to request data from another run. There is no guarantee that such a run has occurred yet, but because all the algorithm and problem instances are already generated during the *prepare* phase, they are already available and data can be generated on them. Only data which is generated during or at the end of the run may miss. As long as the experimenter pay attention to this aspect, he may use the global dataset as wanted.

Coming back to the *produce* phase, the various components executed there need various parameters. Some of them are instantiated directly there, like in your code, but others depend directly on what has been run: `algorithmList` and `problemList`. These two parameters need to be rebuilt from the `dataset` provided, which is easy because the preparation phase gave us the list of algorithm and problem IDs. Because we want only the (algorithm, problem) couples, and not each run, we can only take the instances of a single run of each couple. The most obvious one, which is always present, is the first run (`run = 0`). Thus, for each couple of IDs, we get the context of their first run and retrieve their instances. Because the components need `ExperimentAlgorithm` and `ExperimentProblem`, we also need to retrieve the relevant tags, which are of course available through the datasets (global as well as local). Once the lists are filled, we can use them as parameters for your components, which is what allows us to reuse them as is. Meanwhile, the duplicate filtering has become useless, because we only took one of each. It is worth noting that we could improve it way more by revising each component, but this has nothing to do with the experiment framework anymore, so I let that for other issues.

A last comment is that the components throw `IOException`, so we needed to catch it inside the Lambda. We just throw it again as a `RuntimeException` to remain simple.

# Whole Experiment Revised

The whole code, once it has been revised to use the framework, is the following:
```java
PreparePerformProduceExperimentBuilder<Algorithm<List<BinarySolution>>, BinaryProblem> experimentBuilder = new PreparePerformProduceExperimentBuilder<>();

// Prepare problems
List<ProblemID<? extends BinaryProblem>> problemIDs = new LinkedList<>();
problemIDs.add(experimentBuilder.prepareProblem(() -> new ZDT5()));
problemIDs.add(experimentBuilder.prepareProblem(() -> new OneZeroMax(512)));

// Prepare algorithms
List<AlgorithmID<? extends Algorithm<List<BinarySolution>>>> algorithmIDs = new LinkedList<>();
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new NSGAIIBuilder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new SPEA2Builder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> new MOCellBuilder<BinarySolution>(...) ... .build()));
algorithmIDs.add(experimentBuilder.prepareAlgorithm(
	(problem) -> {
		...
		return new MOCHCBuilder(...) ... .build();
	}));

// Prepare independent runs
experimentBuilder.prepareIndependentRuns(INDEPENDENT_RUNS); // Default: 1

// Ask to run algorithms in parallel
experimentBuilder.performWith(new ThreadPerformer(8)); // Default: SequentialPerformer

// Define run-specific data, extracted from ExecuteAlgorithms, ExperimentAlgorithm, and ExperimentProblem
DataID<String> problemTag = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getProblem().getName());
DataID<String> algorithmTag = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getAlgorithm().getName());
DataID<String> outputDirectoryName = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> experimentBaseDirectory + "/data/" + context.getData(algorithmTag) + "/" + context.getData(problemTag));
DataID<String> funFile = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getData(outputDirectoryName) + "/FUN" + context.getRun() + ".tsv");
DataID<String> varFile = experimentBuilder.defineRunData(Generation.ONCE_AND_BACKUP, (context) -> context.getData(outputDirectoryName) + "/VAR" + context.getRun() + ".tsv");

// Specify pre-/post-run events, extracted from ExecuteAlgorithms, ExperimentAlgorithm, and ExperimentProblem
experimentBuilder.createRunEvent(When.BEFORE_RUN, (context) ->
	JMetalLogger.logger.info(
		"Running algorithm: " + context.getAlgorithm().getName() +
				", problem: " + context.getProblem().getName() +
				", run: " + context.getRun() +
				", funFile: " + context.getData(funFile) +
				", varFile: " + context.getData(varFile)));
experimentBuilder.createRunEvent(When.AFTER_RUN, (context) -> {
	JMetalLogger.logger.info(
		"Finished algorithm: " + context.getAlgorithm().getName() +
				", problem: " + context.getProblem().getName() +
				", run: " + context.getRun());
	new File(context.getData(outputDirectoryName)).mkdirs();
	
	List<BinarySolution> population = context.getAlgorithm().getResult();
	new SolutionListOutput(population)
		.setSeparator("\t")
		.setVarFileOutputContext(new DefaultFileOutputContext(context.getData(varFile)))
		.setFunFileOutputContext(new DefaultFileOutputContext(context.getData(funFile)))
		.print();
	JMetalLogger.logger.info(
		"Generated funFile: " + context.getData(funFile) +
				", varFile: " + context.getData(varFile));
});

// Specify what to do of the raw data with existing components to produce exploitable results
experimentBuilder.produce((dataset) -> {
	try {
		List<ExperimentProblem<BinarySolution>> problemList = new LinkedList<>();
		List<ExperimentAlgorithm<BinarySolution>> algorithmList = new LinkedList<>();
		for (AlgorithmID<? extends Algorithm<List<BinarySolution>>> algorithmID : algorithmIDs) {
			for (ProblemID<? extends BinaryProblem> problemID : problemIDs) {
				RunDataset<? extends Algorithm<List<BinarySolution>>, ? extends BinaryProblem> context = dataset.getRunDataset(algorithmID, problemID, 0);
				problemList.add(new ExperimentProblem<>(context.getProblem(), context.getData(problemTag)));
				algorithmList.add(new ExperimentAlgorithm<>(context.getAlgorithm(), context.getData(algorithmTag), context.getData(problemTag)));
			}
		}
		
		String referenceFrontDirectory = experimentBaseDirectory+"/referenceFronts" ;
		String outputParetoFrontFileName = "FUN" ;
		List<String> referenceFrontFileNames = new LinkedList<>();
		List<GenericIndicator<BinarySolution>> indicatorList = Arrays.asList(
			new Epsilon<BinarySolution>(), new Spread<BinarySolution>(), new GenerationalDistance<BinarySolution>(),
			new PISAHypervolume<BinarySolution>(),
			new InvertedGenerationalDistance<BinarySolution>(),
			new InvertedGenerationalDistancePlus<BinarySolution>()) ;
		String outputParetoSetFileName = "VAR" ;
		String experimentName = "BinaryProblemsStudy";
		
		new GenerateReferenceParetoFront(algorithmList, problemList, experimentBaseDirectory, referenceFrontDirectory, outputParetoFrontFileName, INDEPENDENT_RUNS, referenceFrontFileNames).run();
		new ComputeQualityIndicators<>(algorithmList, problemList, indicatorList, experimentBaseDirectory, referenceFrontDirectory, referenceFrontFileNames, outputParetoFrontFileName, outputParetoSetFileName, INDEPENDENT_RUNS).run();
		new GenerateLatexTablesWithStatistics(algorithmList, problemList, indicatorList, experimentBaseDirectory, experimentName).run();
		new GenerateWilcoxonTestTablesWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
		new GenerateFriedmanTestTables<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).run();
		new GenerateBoxplotsWithR<>(algorithmList, problemList, indicatorList, experimentBaseDirectory).setRows(1).setColumns(2).setDisplayNotch().run();
	} catch (IOException cause) {
		throw new RuntimeException(cause);
	}
});

// Instantiate and run the experiment
Experiment experiment = experimentBuilder.build();
experiment.run();
```

As a summary, the main advantages of this framework so far are:
- the algorithm/problem instantiations are done automatically
- the order of the operations is guaranteed by the builder (excepted inside the *produce* step, which is free, but has been refined to be more constrained by explicit parameters)
- unused data definition is not generated, and otherwise backuped if required

The inconvenients compared to your code:
- it has more intermediaries, especially you need IDs to retrieve instances for algorithms, problems, and data
- it makes explicit the code that you factored in `ExecuteAlgorithm`, so additional support may be implemented to factor this again. It would be good to consider refining and factoring the components too.

**Comment by matthieu-vergne** [comment]
As I said, more can be done. We can imagine additional support for the *produce* phase, in order to cope more easily with producing reference Pareto fronts, computing quality indicators, and other things we may retrieve in most experiments.

**Comment by ajnebro** [comment]
Very nice work :). I would like to try it.

**Comment by matthieu-vergne** [comment]
You can find it on my [experiment branch](https://github.com/matthieu-vergne/jMetal/tree/experiment). The framework has its own module (`jmetal-experiment`) and the rewritten experiment is in the `jmetal-exec` module, with the others. The original one is `BinaryProblemsStudy`, the new one is `BinaryProblemsStudyJEF` (jEF = jMetal Experiment Framework). There is no test, no documentation, only the code. I really focused on the usage, so trying to identify a nice architecture and rewrite the original experiment by adding the relevant methods progressively.

**Comment by matthieu-vergne** [comment]
I added documentation + tests. There is some minor refinements, mainly renaming methods, classes, and generics. Only 1 significant change (which has an impact on the usage), which is to provide a `Supplier` for the performer instead of the instance directly. It was the only thing I needed to change to make experiment instances independent, if you call several times `builder.build()`.

**Comment by ajnebro** [comment]
Ok. I've got your project and I'm taking a look to your implementation. A looooot of work :):)

I will need time to fully understand it, but in the meantime I've been able to run the `BinaryProblemsStudyJEF`. The output of the algorithms seems produce the right results, but the generated Latex files are not correct; I'm attaching two examples:

![binaryproblemstudy](https://user-images.githubusercontent.com/1889196/30387295-53bc12a0-98ac-11e7-8a20-57f3f8f7c36f.png)
![hw_wilcoxon](https://user-images.githubusercontent.com/1889196/30387323-65cdfd28-98ac-11e7-8e85-61e0962a754f.png)

Anyway, these are due surely to minor bugs than can be easily fixed. 

The Prepare step is pretty clear to me, but the Perform one is very complicated. It containts too much details and, thinking in the potential users, I'm affraid that few people is going to understand and use it. Perhaps we could think in encapsulating the, let's say, default behaviour and open the full features to advanced users. The Produce step is also easy to understand.

A minor comment about the names PreparePerformProduce. Perform and Produce are words having a very similar meaning (at least, to me). Perhaps you could consider to use PreparePerformGenerate or PrepareComputeProduce instead.

**Comment by matthieu-vergne** [comment]
> The output of the algorithms seems produce the right results, but the generated Latex files are not correct; I'm attaching two examples:

I don't think I will be able to figure out these ones, at least for now, because it relates to the stuff done in the produce phase, which is mainly your code. I did not go through them yet.

> The Prepare step is pretty clear to me

Good to know.

> but the Perform one is very complicated. It containts too much details and, thinking in the potential users, I'm affraid that few people is going to understand and use it. Perhaps we could think in encapsulating the, let's say, default behavior and open the full features to advanced users.

Well, nothing forbids you to make a factory which provides some default configurations. I have extracted stuff from your classes to make it visible here, so I fairly understand that. But you only need to hide it again.

Actually, it was already planned like that: I did not make it a pull request because it is not finished yet. In my mind, this P3 experiment is only here to satisfy a single goal: isolate the execution of the algorithms as much as possible from the preparation of the experiment and the formatting of the data. I don't consider it as what the final users should actually use. It is only a piece. There is in particular the production phase that I would like to revise, in order to easily plug quality indicators, like we plug algorithms and problems in the preparation phase. But that is the goal of a different class. At the end, we should have several pieces that can be combined in order to form a complete, well managed experiment. And this set of pieces should allow us, maintainers of jMetal, to easily design some typical experiments to provide to our users. And these typical experiments may be provided through a more simple builder, and through a factory for the most simple cases.

We can imagine for example a method taking the list of algorithms, the list of problems, and provide a fully designed experiment with a specific kind of output generated. Such a method would be named depending on the output. A more advanced method may take an additional list of keywords, which identifies the kinds of standard outputs to generate, and so on. With the increasing complexity, people may need instead to use a builder to plug their own quality indicators, for example. And so on.

> A minor comment about the names PreparePerformProduce. Perform and Produce are words having a very similar meaning (at least, to me). Perhaps you could consider to use PreparePerformGenerate or PrepareComputeProduce instead.

I looked for various synonyms to have something we could reduce to a nice acronym. Here, we can reduce it to P3 or 3P, which easily makes it think to "3 steps". People might not remember what are the exact names, but at least they should naturally remember that the first step initializes the experiment, the second step runs algorithms, and the last step creates the outputs. We can change the names, but at least I want to ensure that the resulting acronym makes sense regarding the goal of the class, because this is what we may reuse in future discussions (whatever the full name, it is just too long to call it that way, so the acronym is more important).

**Comment by csgeek95** [comment]
https://github.com/jMetal/jMetal/issues/321#issue-400271753

**Comment by matthieu-vergne** [comment]
> [#321 (comment)](https://github.com/jMetal/jMetal/issues/321#issue-400271753)

This is out of scope. This issue is about the general framework for experiments.

**Comment by natanaelyabes** [comment]
How to perform experiments for the single-objective problem?

**Comment by ajnebro** [comment]
jMetal does not incorporate currently examples of experiments with single-objective optimizers, but it shouldn't be difficult to develop one. Assuming that the result of a single objective metaheuristic is a file containing the best found solution, the idea would be define a quality indicator which would simply return that solution. 

The problem to deal with is that the experiment classes expect that algorithms return a list of solutions, and single objective algorithms return a single value. 

Let me think about a solution to this issue.

**Comment by matthieu-vergne** [comment]
Since it is again out of scope, could you please rename this issue as
"Redesign of experiments in jMetal"?
I hope it would reduce the chance for people to ask for anything about
experiments here instead of opening their own issue.
We just can't let questions and answers be lost in a general issue,
otherwise people won't know if a question has already been answered.

Le mar. 19 nov. 2019 à 18:33, Antonio J. Nebro <notifications@github.com> a
écrit :
